python/lib/adaguc/CGIRunner.py

```diff
@@ -37,8 +37,14 @@ def run(self, cmds, url, output, env=[],  path=None, isCGI=True):
         # Execute adaguc-server binary
         ON_POSIX = 'posix' in sys.builtin_module_names
         process = Popen(cmds, stdout=PIPE, stderr=PIPE, env=localenv, close_fds=ON_POSIX)
-        (processOutput, processError) = process.communicate()
- 
+        
+        try:
+            (processOutput, processError) = process.communicate(timeout=timeout)
+        except TimeoutExpired:
+            process.kill()
```

Comment:
According to [this](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate), you also need to do a `proc.communicate()` to do a proper cleanup?

Reply:
yes, good one. Clean kill.